components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -2,89 +2,2001 @@
+        {
+            self.pending_pull_intos
+                .borrow_mut()
+                .push(pull_into_descriptor);
+        }
+
+        // Perform ! ReadableStreamAddReadIntoRequest(stream, readIntoRequest).
+        stream.add_read_into_request(read_into_request);
+
+        // Perform ! ReadableByteStreamControllerCallPullIfNeeded(controller).
+        self.call_pull_if_needed(can_gc);
     }
 
     /// <https://streams.spec.whatwg.org/#readable-byte-stream-controller-respond>
-    pub(crate) fn respond(&self, _bytes_written: u64) -> Fallible<()> {
-        todo!()
+    #[allow(unsafe_code)]
+    pub(crate) fn respond(&self, bytes_written: u64, can_gc: CanGc) -> Fallible<()> {
+        let cx = GlobalScope::get_cx();
+        {
+            // Assert: controller.[[pendingPullIntos]] is not empty.
+            let mut pending_pull_intos = self.pending_pull_intos.borrow_mut();
+            assert!(!pending_pull_intos.is_empty());
+
+            // Let firstDescriptor be controller.[[pendingPullIntos]][0].
+            let first_descriptor = pending_pull_intos.first_mut().unwrap();
+
+            // Let state be controller.[[stream]].[[state]].
+            let stream = self.stream.get().unwrap();
+
+            // If state is "closed",
+            // If bytesWritten is not 0, throw a TypeError exception.
+            if stream.is_closed() && bytes_written != 0 {
+                return Err(Error::Type("bytesWritten is not 0".to_owned()));
+            }
+
+            // Assert: state is "readable".
+            assert!(stream.is_readable());
+
+            // If bytesWritten is 0, throw a TypeError exception.
+            if bytes_written == 0 {
+                return Err(Error::Type("bytesWritten is 0".to_owned()));
+            }
+
+            // If firstDescriptor’s bytes filled + bytesWritten > firstDescriptor’s byte length,
+            // throw a RangeError exception.
+            if first_descriptor.bytes_filled.get() + bytes_written > first_descriptor.byte_length {
+                return Err(Error::Range(
+                    "bytes filled + bytesWritten > byte length".to_owned(),
+                ));
+            }
+
+            // Set firstDescriptor’s buffer to ! TransferArrayBuffer(firstDescriptor’s buffer).
+            first_descriptor.buffer = first_descriptor
+                .buffer
+                .transfer_array_buffer(cx)
+                .ok_or_else(|| {
+                    rooted!(in(*cx) let mut rval = UndefinedValue());
+                    unsafe {
+                        assert!(JS_GetPendingException(*cx, rval.handle_mut()));
+                        JS_ClearPendingException(*cx)
+                    };
+
+                    Error::Type("can't transfer array buffer".to_owned())
+                })?;
+        }
+
+        // Perform ? ReadableByteStreamControllerRespondInternal(controller, bytesWritten).
+        self.respond_internal(bytes_written, can_gc);
```

Comment:
So there is something I am only now starting to appreciate, which is that `? Op` means the equivalent of `Op()?;` in rust. So `respond_internal` should be `Fallible`, and you should unwrap that call here. 

And `! Op` is the same as `Op().unwrap()` in Rust(meaning it should never actually fail). 

See https://tc39.es/ecma262/multipage/notational-conventions.html#sec-returnifabrupt-shorthands

Just highlighting this now, clean-up could be done in a follow-up. But if you have a test that fails and you don't know why,it might be because of that sometimes.

Reply:
For example, `enqueue_cloned_chunk_to_queue` should actually return a result, see https://github.com/servo/servo/blob/490cd3f99cf18af0e28c43237a8f788d65dfefa7/components/script/dom/readablebytestreamcontroller.rs#L1464